fastfield_codecs/src/serialize.rs

```diff
@@ -0,0 +1,245 @@
+) -> io::Result<()> {
+    let column = monotonic_map_column(typed_column, T::to_u64);
+    let gcd_params_opt = if codecs.contains(&FastFieldCodecType::Gcd) {
+        gcd_params(&column)
+    } else {
+        None
+    };
+
+    let gcd_params = if let Some(gcd_params) = gcd_params_opt {
+        gcd_params
+    } else {
+        return serialize_without_gcd(column, output, codecs);
+    };
+
+    write_header(FastFieldCodecType::Gcd, output)?;
+    let base_value = column.min_value();
+    let gcd_divider = DividerU64::divide_by(gcd_params.gcd);
+    let divided_fastfield_accessor =
+        monotonic_map_column(column, |val: u64| gcd_divider.divide(val - base_value));
+
+    serialize_without_gcd(divided_fastfield_accessor, output, codecs)?;
+
+    gcd_params.serialize(output)?;
+    Ok(())
+}
+
+fn serialize_without_gcd(
+    column: impl Column<u64>,
+    output: &mut impl io::Write,
+    codecs: &[FastFieldCodecType],
+) -> io::Result<()> {
+    let mut estimations = Vec::new();
+    for &codec in codecs {
+        if codec == FastFieldCodecType::Gcd {
+            continue;
+        }
+        match codec {
+            FastFieldCodecType::Bitpacked => {
+                codec_estimation::<BitpackedCodec, _>(&column, &mut estimations);
+            }
+            FastFieldCodecType::Linear => {
+                codec_estimation::<LinearCodec, _>(&column, &mut estimations);
+            }
+            FastFieldCodecType::BlockwiseLinear => {
+                codec_estimation::<BlockwiseLinearCodec, _>(&column, &mut estimations);
+            }
+            FastFieldCodecType::Gcd => {}
+        }
+    }
+    if let Some(broken_estimation) = estimations.iter().find(|estimation| estimation.0.is_nan()) {
+        warn!(
+            "broken estimation for fast field codec {:?}",
+            broken_estimation.1
+        );
+    }
+    // removing nan values for codecs with broken calculations, and max values which disables
+    // codecs
+    estimations.retain(|estimation| !estimation.0.is_nan() && estimation.0 != f32::MAX);
+    estimations.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
+    let (_ratio, codec_type) = estimations[0];
+
+    write_header(codec_type, output)?;
+    match codec_type {
+        FastFieldCodecType::Bitpacked => {
+            BitpackedCodec::serialize(output, &column)?;
+        }
+        FastFieldCodecType::Linear => {
+            LinearCodec::serialize(output, &column)?;
+        }
+        FastFieldCodecType::BlockwiseLinear => {
+            BlockwiseLinearCodec::serialize(output, &column)?;
+        }
+        FastFieldCodecType::Gcd => {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "GCD codec not supported.",
+            ));
+        }
+    }
+    output.flush()?;
+    Ok(())
+}
+
+pub fn serialize_and_load<T: MonotonicallyMappableToU64 + Ord + Default>(
+    column: &[T],
+) -> Arc<dyn Column<T>> {
+    let mut buffer = Vec::new();
```

Comment:
call `serialize_and_load_given_codecs` with ALL_CODEC_TYPES

Reply:
ah right. Actuall serialize_and_load_given_codecs is not used so I removed it.